### DISCUSSION TITLE
Model-view-controller for egui - how?

### DISCUSSION THREAD

**Discussion by John-Nagle** [emilk/egui#1806] [Q&A]
Are there any libraries to help with model-view-controller architecture with **egui**? Basically, there you are in the middle of refresh, someone has clicked on something, and now you need to send a message to something to deal with the event. Later, you'll get an answer back, probably on a later frame, and that answer needs to be routed to the persistent part of the appropriate GUI element. That's pretty normal.

So is there some crate to help manage that? Something that owns the persistent part of the GUI elements and has the appropriate message passing? This has to have been done before.

**Comment by camelid** [comment]
I'm also curious about this, and in general what design patterns or architectures are recommended for egui.

**Reply by John-Nagle** [reply]
I never got an answer, so I wrote one. See https://github.com/John-Nagle/ui-mock
It's mostly for my own use, but it works for what I use it for.
